conans/client/cmd/uploader.py

```diff
@@ -246,21 +263,48 @@ def upload_package(self, pref, prev_bundle, remote):
         self._output.debug(f"Upload {pref} in {duration} time")
 
 
-def compress_files(files, name, dest_dir, compresslevel=None, ref=None):
+def compress_files(files, name, dest_dir, compressformat=None, compresslevel=None, ref=None):
     t1 = time.time()
-    # FIXME, better write to disk sequentially and not keep tgz contents in memory
-    tgz_path = os.path.join(dest_dir, name)
-    if name in (PACKAGE_TGZ_NAME, EXPORT_SOURCES_TGZ_NAME) and len(files) > 100:
-        ref_name = f"{ref}:" or ""
+    tar_path = os.path.join(dest_dir, name)
+    if name in (PACKAGE_TGZ_NAME, PACKAGE_TZSTD_NAME, EXPORT_SOURCES_TGZ_NAME) and len(files) > 100:
+        ref_name = f"{ref}:" if ref else ""
         ConanOutput().info(f"Compressing {ref_name}{name}")
-    with set_dirty_context_manager(tgz_path), open(tgz_path, "wb") as tgz_handle:
-        tgz = gzopen_without_timestamps(name, mode="w", fileobj=tgz_handle,
-                                        compresslevel=compresslevel)
-        for filename, abs_path in sorted(files.items()):
-            # recursive is False in case it is a symlink to a folder
-            tgz.add(abs_path, filename, recursive=False)
-        tgz.close()
+
+    if compressformat == "zstd":
+        with open(tar_path, "wb") as tarfile_obj:
+            def reset_tarinfo(tarinfo):
+                """
+                Resets mtime in the tarinfo for consistency with
+                gzopen_without_timestamps()
+                """
+                tarinfo.mtime = 0
+                return tarinfo
+
+            # Only provide level if it was overridden by config.
+            zstd_kwargs = {}
+            if compresslevel is not None:
+                zstd_kwargs["level"] = compresslevel
+
+            dctx = zstandard.ZstdCompressor(**zstd_kwargs)
+
+            # Create a zstd stream writer so tarfile writes uncompressed data to
+            # the zstd stream writer, which in turn writes compressed data to the
+            # output tar.zst file.
+            with dctx.stream_writer(tarfile_obj) as stream_writer:
+                with tarfile.open(mode="w|", fileobj=stream_writer,
+                                  format=tarfile.PAX_FORMAT) as tar:
+                    for filename, abs_path in sorted(files.items()):
+                        tar.add(abs_path, filename, recursive=False, filter=reset_tarinfo)
```

Comment:
You will most likely want to issue an `flush(zstandard.COMPRESSOBJ_FLUSH_BLOCK)` whenever the written file size exceeds a couple of megabytes.

This has no relevance for Conan as a client, as it's only accessing the compressed file as a stream, but it is very much relevant for Artifactory and alike as they can build on that to properly ***index*** into the .tar.gz, as de-compression can be restartet after every frame.

It's still entirely compatible with the tar / zstd command line tools.